README.md
![Language](https://img.shields.io/badge/English-brigthgreen)

# Motion Detector

![PyPI](https://img.shields.io/pypi/v/motion-detector-g4)
![PyPI - License](https://img.shields.io/pypi/l/motion-detector-g4)
![PyPI - Python Version](https://img.shields.io/pypi/pyversions/motion-detector-g4)

Python module for comparing two frames from a video for motion.

***

## Installation

### Package Installation from PyPi

```bash
$ pip install motion-detector-g4
```

### Package Installation from Source Code

The source code is available on [GitHub](https://github.com/Genzo4/motion_detector).  
Download and install the package:

```bash
$ git clone https://github.com/Genzo4/motion_detector
$ cd motion_detector
$ pip install -r requirements.txt
$ pip install .
```

***

## Basic usage

Import:
```python
from motion_detector_g4 import MotionDetector
```

Create an instance of the motion detector. You can specify additional options:
- min_area - minimum tracked change size (blob size).
  Default value: 4000.
- max_area - maximum tracked change size (blob size).
  Default value: 150000.
- noise_size - the maximum size of the removed noise.
  Default value: 10.
- debug - debug mode. If it is enabled, intermediate frames are created, 
  showing the process of processing.
  Default value: False.

```python
md = MotionDetector(min_area=4000, max_area=150000, noise_size=10, debug=False)
```

The module uses the following algorithm:
1. The first frame is being processed (method apply_first_frame).
![Img. 1](https://github.com/Genzo4/motion_detector/raw/main/images/01_frame_1.png "Img. 1 - First frame")
2. The next frame is being processed (method check_motion):
![Img. 2](https://github.com/Genzo4/motion_detector/raw/main/images/01_frame_2.png "Img. 2 - Next frame")
   - the first frame is subtracted from this frame using the module BackgroundSubtractorMOG2
     from the library OpenCV. If debug mode is enabled, then a file is created with an additional 
     extension "mask" with the result of the module.  
     ![Img. 3](https://github.com/Genzo4/motion_detector/raw/main/images/01_frame_2.mask.png "Img. 3 - Removed the background")
   - remove noise using the morphologyEx module from the OpenCV library. This process 
     configured using the noise_size parameter. If debug mode is enabled, then 
     a file is created with the additional extension "clear" with the result of the module's operation.
     ![Img. 4](https://github.com/Genzo4/motion_detector/raw/main/images/01_frame_2.clear.png "Img. 4 - Remove noise")
   - looking for areas of motion (blobs) larger than min_area but smaller than max_area. If debug mode is enabled, then 
     files are created with the extension "blobs" and "blobs2" with the result of the module's operation 
     (found areas are circled in red).
     ![Img. 5](https://github.com/Genzo4/motion_detector/raw/main/images/01_frame_2.blobs.png "Img. 5 - Found areas of interest")
     ![Img. 6](https://github.com/Genzo4/motion_detector/raw/main/images/01_frame_2.blobs2.png "Img. 6 - Found areas of interest")

See the example.py file for an example of usage.

***

![Language](https://img.shields.io/badge/Русский-brigthgreen)

# Motion Detector

![PyPI](https://img.shields.io/pypi/v/motion-detector-g4)
![PyPI - License](https://img.shields.io/pypi/l/motion-detector-g4)
![PyPI - Python Version](https://img.shields.io/pypi/pyversions/motion-detector-g4)

Python модуль для сравнения двух кадров из видеозаписи на предмет наличия движения в них.

***

## Установка

### Установка пакета с PyPi

```bash
$ pip install motion-detector-g4
```

### Установка пакета из исходного кода

Исходный код размещается на [GitHub](https://github.com/Genzo4/motion_detector).  
Скачайте его и установите пакет:

```bash
$ git clone https://github.com/Genzo4/motion_detector
$ cd motion_detector
$ pip install -r requirements.txt
$ pip install .
```

***

## Использование

Подключаем:
```python
from motion_detector_g4 import MotionDetector
```

Создаём экземпляр детектора движения. Можно указать дополнительные параметры:
- min_area - минимальный размер отслеживаемого изменения (размера blob'а).
  Значение по умолчанию: 4000.
- max_area - максимальный размер отслеживаемого изменения (размера blob'а).
  Значение по умолчанию: 150000.
- noise_size - максимальный размер удаляемого "шума".
  Значение по умолчанию: 10.
- debug - режим отладки. Если его включить, то создаются промежуточные кадры,
  показывающие процесс обработки.
  Значение по умолчанию: False.

```python
md = MotionDetector(min_area=4000, max_area=150000, noise_size=10, debug=False)
```

В модуле используется следующий алгоритм:
1. Подаётся на обработку первый кадр (метод apply_first_frame).
![Рис. 1](https://github.com/Genzo4/motion_detector/raw/main/images/01_frame_1.png "Рис. 1 - Первый кадр")
2. Подаётся на обработку следующий кадр (метод check_motion):
![Рис. 2](https://github.com/Genzo4/motion_detector/raw/main/images/01_frame_2.png "Рис. 2 - Следующий кадр")
   - из этого кадра "вычитается" первый кадр с помощью модуля BackgroundSubtractorMOG2
     из библиотеки OpenCV. Если включён режим отладки, то создаётся файл с добавочным
     расширением "mask" с результатом работы модуля.  
     ![Рис. 3](https://github.com/Genzo4/motion_detector/raw/main/images/01_frame_2.mask.png "Рис. 3 - Убрали фон")
   - удаляем "шум" с помощью модуля morphologyEx из библиотеки OpenCV. Данный процесс
     настраивается с помощью параметра noise_size. Если включён режим отладки, то
     создаётся файл с добавочным расширением "clear" с результатом работы модуля.
     ![Рис. 4](https://github.com/Genzo4/motion_detector/raw/main/images/01_frame_2.clear.png "Рис. 4 - Очистили от шума")
   - ищем области движения (blob'ы) размером больше min_area, но меньше max_area. Если включён режим отладки, то
     создаются файлы с добавочным расширением "blobs" и "blobs2" с результатом работы модуля 
     (найденные области обводятся красными кругами).
     ![Рис. 5](https://github.com/Genzo4/motion_detector/raw/main/images/01_frame_2.blobs.png "Рис. 5 - Нашли интересующие области движения")
     ![Рис. 6](https://github.com/Genzo4/motion_detector/raw/main/images/01_frame_2.blobs2.png "Рис. 6 - Нашли интересующие области движения")

Пример использования см. в файле example.py


CHANGELOG.md
# Motion Detector change log

This document records all notable changes to
[Motion Detector](https://github.com/Genzo4/motion_detector).

## 1.2.0 (2023-10-05)

- Work with frames (numpy.ndarray)

## 1.1.0 (2022-09-01)

- Change names to PEP 8

## 1.0.0 (2022-06-02)

- First public release


tests/test_addExt.py
import pytest
from motion_detector_g4 import MotionDetector


def test_1():
    path = 'c:\\test\\test.png'
    ext = 'ext'
    newPath = MotionDetector._addExt(path, ext)
    assert newPath == 'c:\\test\\test.ext.png'


def test_2():
    path = '/test/test.png'
    ext = 'ext'
    newPath = MotionDetector._addExt(path, ext)
    assert newPath == '/test/test.ext.png'


def test_3():
    path = 'test.png'
    ext = 'ext'
    newPath = MotionDetector._addExt(path, ext)
    assert newPath == 'test.ext.png'


def test_4():
    path = 'c:\\test\\test'
    ext = 'ext'
    newPath = MotionDetector._addExt(path, ext)
    assert newPath == 'c:\\test\\test.ext'


def test_5():
    path = 'c:\\test\\test.2.png'
    ext = 'ext'
    newPath = MotionDetector._addExt(path, ext)
    assert newPath == 'c:\\test\\test.2.ext.png'



tests/test_debugMod.py
import pytest
from motion_detector_g4 import MotionDetector
import os
import glob
from itertools import chain
import cv2
from utilspy_g4 import templatedRemoveFiles


def _removeTempFiles() -> None:
    """
    Remove temp files
    :rtype: None
    :return: None
    """

    templatedRemoveFiles('tests/images/*.clear.*')
    templatedRemoveFiles('tests/images/*.mask.*')
    templatedRemoveFiles('tests/images/*.blobs.*')
    templatedRemoveFiles('tests/images/*.blobs2.*')


def test_1():
    _removeTempFiles()

    md = MotionDetector()
    md.applyFirstFrame('tests/images/01_frame_1.png')
    md.checkMotion('tests/images/01_frame_2.png')
    assert os.path.exists('tests/images/01_frame_2.clear.png') == False
    assert os.path.exists('tests/images/01_frame_2.mask.png') == False
    assert os.path.exists('tests/images/01_frame_2.blobs.png') == False
    assert os.path.exists('tests/images/01_frame_2.blobs2.png') == False
    _removeTempFiles()


def test_2():
    md = MotionDetector(debug=True)
    md.applyFirstFrame('tests/images/01_frame_1.png')
    md.checkMotion('tests/images/01_frame_2.png')
    assert os.path.exists('tests/images/01_frame_2.clear.png') == True
    assert os.path.exists('tests/images/01_frame_2.mask.png') == True
    assert os.path.exists('tests/images/01_frame_2.blobs.png') == False
    assert os.path.exists('tests/images/01_frame_2.blobs2.png') == False
    _removeTempFiles()


def test_3():
    md = MotionDetector(minArea=2000, debug=True)
    md.applyFirstFrame('tests/images/01_frame_1.png')
    md.checkMotion('tests/images/01_frame_2.png')
    assert os.path.exists('tests/images/01_frame_2.clear.png') == True
    assert os.path.exists('tests/images/01_frame_2.mask.png') == True
    assert os.path.exists('tests/images/01_frame_2.blobs.png') == True
    assert os.path.exists('tests/images/01_frame_2.blobs2.png') == True
    _removeTempFiles()


def test_mask():
    md = MotionDetector(debug=True)
    md.applyFirstFrame('tests/images/01_frame_1.png')
    md.checkMotion('tests/images/01_frame_2.png')

    frameMask = cv2.imread('tests/images/01_frame_2.mask.png')

    (b, g, r) = frameMask[17, 65]
    assert int(r) + int(g) + int(b) == 765

    (b, g, r) = frameMask[38, 36]
    assert int(r) + int(g) + int(b) == 0

    (b, g, r) = frameMask[104, 1713]
    assert int(r) + int(g) + int(b) == 765

    (b, g, r) = frameMask[53, 1752]
    assert int(r) + int(g) + int(b) == 0

    (b, g, r) = frameMask[881, 1868]
    assert int(r) + int(g) + int(b) == 765

    (b, g, r) = frameMask[877, 1834]
    assert int(r) + int(g) + int(b) == 0

    (b, g, r) = frameMask[1002, 62]
    assert int(r) + int(g) + int(b) == 765

    (b, g, r) = frameMask[994, 140]
    assert int(r) + int(g) + int(b) == 0

    _removeTempFiles()


def test_clear():
    md = MotionDetector(debug=True)
    md.applyFirstFrame('tests/images/01_frame_1.png')
    md.checkMotion('tests/images/01_frame_2.png')

    frameClear = cv2.imread('tests/images/01_frame_2.clear.png')

    (b, g, r) = frameClear[17, 65]
    assert int(r) + int(g) + int(b) == 0

    (b, g, r) = frameClear[38, 36]
    assert int(r) + int(g) + int(b) == 0

    (b, g, r) = frameClear[104, 1713]
    assert int(r) + int(g) + int(b) == 0

    (b, g, r) = frameClear[53, 1752]
    assert int(r) + int(g) + int(b) == 0

    (b, g, r) = frameClear[881, 1868]
    assert int(r) + int(g) + int(b) == 765

    (b, g, r) = frameClear[877, 1834]
    assert int(r) + int(g) + int(b) == 0

    (b, g, r) = frameClear[1002, 62]
    assert int(r) + int(g) + int(b) == 0

    (b, g, r) = frameClear[994, 140]
    assert int(r) + int(g) + int(b) == 0

    _removeTempFiles()


tests/test_debug_mod.py
import pytest
import os
import cv2

from motion_detector_g4 import MotionDetector
from utilspy_g4 import templated_remove_files


def _remove_temp_files() -> None:
    """
    Remove temp files
    :rtype: None
    :return: None
    """

    templated_remove_files('tests/images/*.clear.*')
    templated_remove_files('tests/images/*.mask.*')
    templated_remove_files('tests/images/*.blobs.*')
    templated_remove_files('tests/images/*.blobs2.*')


def test_1():
    _remove_temp_files()

    md = MotionDetector()
    md.apply_first_frame('tests/images/01_frame_1.png')
    md.check_motion('tests/images/01_frame_2.png')
    assert os.path.exists('tests/images/01_frame_2.clear.png') is False
    assert os.path.exists('tests/images/01_frame_2.mask.png') is False
    assert os.path.exists('tests/images/01_frame_2.blobs.png') is False
    assert os.path.exists('tests/images/01_frame_2.blobs2.png') is False
    _remove_temp_files()


def test_2():
    md = MotionDetector(debug=True)
    md.apply_first_frame('tests/images/01_frame_1.png')
    md.check_motion('tests/images/01_frame_2.png')
    assert os.path.exists('tests/images/01_frame_2.clear.png') is True
    assert os.path.exists('tests/images/01_frame_2.mask.png') is True
    assert os.path.exists('tests/images/01_frame_2.blobs.png') is False
    assert os.path.exists('tests/images/01_frame_2.blobs2.png') is False
    _remove_temp_files()


def test_3():
    md = MotionDetector(min_area=2000, debug=True)
    md.apply_first_frame('tests/images/01_frame_1.png')
    md.check_motion('tests/images/01_frame_2.png')
    assert os.path.exists('tests/images/01_frame_2.clear.png') is True
    assert os.path.exists('tests/images/01_frame_2.mask.png') is True
    assert os.path.exists('tests/images/01_frame_2.blobs.png') is True
    assert os.path.exists('tests/images/01_frame_2.blobs2.png') is True
    _remove_temp_files()


def test_mask():
    md = MotionDetector(debug=True)
    md.apply_first_frame('tests/images/01_frame_1.png')
    md.check_motion('tests/images/01_frame_2.png')

    frame_mask = cv2.imread('tests/images/01_frame_2.mask.png')

    (b, g, r) = frame_mask[17, 65]
    assert int(r) + int(g) + int(b) == 765

    (b, g, r) = frame_mask[38, 36]
    assert int(r) + int(g) + int(b) == 0

    (b, g, r) = frame_mask[104, 1713]
    assert int(r) + int(g) + int(b) == 765

    (b, g, r) = frame_mask[53, 1752]
    assert int(r) + int(g) + int(b) == 0

    (b, g, r) = frame_mask[881, 1868]
    assert int(r) + int(g) + int(b) == 765

    (b, g, r) = frame_mask[877, 1834]
    assert int(r) + int(g) + int(b) == 0

    (b, g, r) = frame_mask[1002, 62]
    assert int(r) + int(g) + int(b) == 765

    (b, g, r) = frame_mask[994, 140]
    assert int(r) + int(g) + int(b) == 0

    _remove_temp_files()


def test_clear():
    md = MotionDetector(debug=True)
    md.apply_first_frame('tests/images/01_frame_1.png')
    md.check_motion('tests/images/01_frame_2.png')

    frame_clear = cv2.imread('tests/images/01_frame_2.clear.png')

    (b, g, r) = frame_clear[17, 65]
    assert int(r) + int(g) + int(b) == 0

    (b, g, r) = frame_clear[38, 36]
    assert int(r) + int(g) + int(b) == 0

    (b, g, r) = frame_clear[104, 1713]
    assert int(r) + int(g) + int(b) == 0

    (b, g, r) = frame_clear[53, 1752]
    assert int(r) + int(g) + int(b) == 0

    (b, g, r) = frame_clear[881, 1868]
    assert int(r) + int(g) + int(b) == 765

    (b, g, r) = frame_clear[877, 1834]
    assert int(r) + int(g) + int(b) == 0

    (b, g, r) = frame_clear[1002, 62]
    assert int(r) + int(g) + int(b) == 0

    (b, g, r) = frame_clear[994, 140]
    assert int(r) + int(g) + int(b) == 0

    _remove_temp_files()


motion_detector_g4/__init__.py
from cv2 import cv2, SimpleBlobDetector
import numpy
from utilspy_g4 import add_ext
from functools import singledispatchmethod


class MotionDetector:

    def __init__(self, min_area: int = 4000, max_area: int = 150000, noise_size: int = 10, debug: bool = False):
        """
        :param min_area: Min blob size
        :param max_area: Max blob size
        :param noise_size: Max size of noise area
        :param debug: Is debug mod
        """

        self.debug = debug

        self.back_sub = cv2.createBackgroundSubtractorMOG2(history=1, detectShadows=False)

        self.denoise_kernel = numpy.ones((noise_size, noise_size), numpy.uint8)

        self.blob_detector = self._create_blob_detector(min_area, max_area)

    @staticmethod
    def _create_blob_detector(min_area: int, max_area: int) -> SimpleBlobDetector:
        """
        Create and config OpenCV Simple Blob Detector

        :param min_area: Min blob size
        :param max_area: Max blob size
        :rtype: SimpleBlobDetector
        :return: SimpleBlobDetector
        """

        params = cv2.SimpleBlobDetector_Params()

        params.filterByColor = False

        params.minRepeatability = 1
        params.minThreshold = 250
        params.maxThreshold = 255

        # Filter by Area
        params.filterByArea = True
        params.minArea = min_area
        params.maxArea = max_area

        params.filterByCircularity = False
        params.filterByConvexity = False
        params.filterByInertia = False

        return cv2.SimpleBlobDetector_create(params)

    @singledispatchmethod
    def apply_first_frame(self, first_frame) -> None:
        raise NotImplementedError(f"Cannot format value of type {type(first_frame)}")

    @apply_first_frame.register
    def _(self, first_frame_path: str) -> None:
        """
        :param first_frame_path:
        :rtype: None
        :return: None
        """

        first_frame = cv2.imread(first_frame_path)

        self.apply_first_frame(first_frame)

    @apply_first_frame.register
    def _(self, first_frame: numpy.ndarray) -> None:
        """
        :param first_frame:
        :rtype: None
        :return: None
        """
        self.back_sub.apply(first_frame)

    @singledispatchmethod
    def check_motion(self, next_frame) -> bool:
        raise NotImplementedError(f"Cannot format value of type {type(next_frame)}")

    @check_motion.register
    def _(self, next_frame_path: str) -> bool:
        """
        :param next_frame_path: Next frame for comparison
        :rtype: bool
        :return: Is motion
        """

        next_frame = cv2.imread(next_frame_path)

        # 1. Delete background

        frame_mask = self.back_sub.apply(next_frame)

        if self.debug:
            cv2.imwrite(add_ext(next_frame_path, 'mask'), frame_mask)

        # 2. Clear noises

        frame_clear = cv2.morphologyEx(frame_mask, cv2.MORPH_OPEN, self.denoise_kernel)

        if self.debug:
            cv2.imwrite(add_ext(next_frame_path, 'clear'), frame_clear)

        # 3. Search blobs

        blobs = self.blob_detector.detect(frame_clear)

        if len(blobs) > 0:
            if self.debug:
                frame_with_blobs = cv2.drawKeypoints(next_frame, blobs, numpy.array([]), (0, 0, 255),
                                                     cv2.DRAW_MATCHES_FLAGS_DRAW_RICH_KEYPOINTS)
                frame_mask_with_blobs = cv2.drawKeypoints(frame_clear, blobs, numpy.array([]), (0, 0, 255),
                                                          cv2.DRAW_MATCHES_FLAGS_DRAW_RICH_KEYPOINTS)

                cv2.imwrite(add_ext(next_frame_path, 'blobs'), frame_with_blobs)
                cv2.imwrite(add_ext(next_frame_path, 'blobs2'), frame_mask_with_blobs)

            return True

        return False

    @check_motion.register
    def _(self, next_frame: numpy.ndarray) -> bool:

        # 1. Delete background

        frame_mask = self.back_sub.apply(next_frame)

        # 2. Clear noises

        frame_clear = cv2.morphologyEx(frame_mask, cv2.MORPH_OPEN, self.denoise_kernel)

        # 3. Search blobs

        blobs = self.blob_detector.detect(frame_clear)

        if len(blobs) > 0:
            return True

        return False
